File: java/starter/template/src/main/java/network/t0/provider/internal/PublishPaymentIntentQuotes.java
package network.t0.provider.internal;

import com.google.protobuf.Timestamp;
import network.t0.sdk.proto.tzero.v1.common.Decimal;
import network.t0.sdk.proto.tzero.v1.common.PaymentMethodType;
import network.t0.sdk.proto.tzero.v1.payment_intent.PaymentIntentServiceGrpc;
import network.t0.sdk.proto.tzero.v1.payment_intent.UpdateQuoteRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.UUID;

/**
 * Pay-In Provider role — Step 3A.1.
 *
 * <p>Publishes pay-in quotes for the payment intent flow. These quotes tell the
 * network what exchange rates you're willing to accept when an end-user pays via
 * one of your supported payment methods.
 *
 * <p>TODO: Step 3A.1 replace this with fetching pay-in quotes from your systems
 * and publishing them into t-0 Network. We recommend publishing at least once per
 * 5 seconds, but not more than once per second.
 */
public class PublishPaymentIntentQuotes {

    private static final Logger log = LoggerFactory.getLogger(PublishPaymentIntentQuotes.class);

    public static void publish(PaymentIntentServiceGrpc.PaymentIntentServiceBlockingStub paymentIntentClient) {
        try {
            Instant now = Instant.now();
            Timestamp expiration = Timestamp.newBuilder()
                    .setSeconds(now.plusSeconds(30).getEpochSecond())
                    .setNanos(now.plusSeconds(30).getNano())
                    .build();
            Timestamp timestamp = Timestamp.newBuilder()
                    .setSeconds(now.getEpochSecond())
                    .setNanos(now.getNano())
                    .build();

            // NOTE: Every UpdateQuote request discards all previous payment intent quotes
            // that were published before. Combine multiple quotes into a single request.
            paymentIntentClient.updateQuote(UpdateQuoteRequest.newBuilder()
                    .addPaymentIntentQuotes(UpdateQuoteRequest.Quote.newBuilder()
                            .setCurrency("EUR")
                            .setPaymentMethod(PaymentMethodType.PAYMENT_METHOD_TYPE_SEPA)
                            .setExpiration(expiration)
                            .setTimestamp(timestamp)
                            .addBands(UpdateQuoteRequest.Quote.Band.newBuilder()
                                    .setClientQuoteId(UUID.randomUUID().toString())
                                    .setMaxAmount(Decimal.newBuilder()
                                            .setUnscaled(1000) // max 1000 USD for this band
                                            .setExponent(0)
                                            .build())
                                    // rate is always USD/XXX, so for EUR quote should be USD/EUR
                                    .setRate(Decimal.newBuilder()
                                            .setUnscaled(92) // rate 0.92
                                            .setExponent(-2)
                                            .build())
                                    .build())
                            .build())
                    .build());

            log.debug("✅ Payment intent quotes published successfully");

        } catch (io.grpc.StatusRuntimeException e) {
            log.error("Error updating payment intent quote: {} - {}", e.getStatus().getCode(), e.getMessage());
        }
    }
}

File: python/starter/src/t0_provider_starter/template/src/provider/publish_payment_intent_quotes.py
"""Payment intent quote publishing to the T-0 Network.

Go equivalent: internal/publish_payment_intent_quotes.go → PublishPaymentIntentQuotes()

Pay-In Provider role — Step 3A.1. Publishes sample pay-in quotes every 5 seconds.
Replace the hardcoded values with quotes from your own pricing systems.
"""

from __future__ import annotations

import asyncio
import logging
import uuid

from google.protobuf.timestamp_pb2 import Timestamp
from t0_provider_sdk.api.tzero.v1.common.common_pb2 import Decimal
from t0_provider_sdk.api.tzero.v1.common.payment_method_pb2 import PAYMENT_METHOD_TYPE_SEPA
from t0_provider_sdk.api.tzero.v1.payment_intent.network_connect import PaymentIntentServiceClient
from t0_provider_sdk.api.tzero.v1.payment_intent.network_pb2 import UpdateQuoteRequest

logger = logging.getLogger(__name__)

PUBLISH_INTERVAL_SECONDS = 5


def _now_timestamp() -> Timestamp:
    ts = Timestamp()
    ts.GetCurrentTime()
    return ts


def _expiration_timestamp(seconds_from_now: int = 30) -> Timestamp:
    ts = _now_timestamp()
    ts.seconds += seconds_from_now
    return ts


async def publish_payment_intent_quotes(
    payment_intent_client: PaymentIntentServiceClient,
    shutdown_event: asyncio.Event,
) -> None:
    """Publish sample pay-in quotes on a regular interval.

    NOTE: Every UpdateQuote request discards all previous payment intent quotes.
    If you want to publish multiple quotes, combine them into a single request.
    """
    while not shutdown_event.is_set():
        try:
            await payment_intent_client.update_quote(
                UpdateQuoteRequest(
                    payment_intent_quotes=[
                        UpdateQuoteRequest.Quote(
                            currency="EUR",
                            payment_method=PAYMENT_METHOD_TYPE_SEPA,
                            expiration=_expiration_timestamp(30),
                            timestamp=_now_timestamp(),
                            bands=[
                                UpdateQuoteRequest.Quote.Band(
                                    client_quote_id=str(uuid.uuid4()),
                                    max_amount=Decimal(unscaled=1000, exponent=0),  # max 1000 USD
                                    # rate is always USD/XXX, so for EUR quote should be USD/EUR
                                    rate=Decimal(unscaled=92, exponent=-2),  # rate 0.92
                                ),
                            ],
                        ),
                    ],
                ),
            )
        except Exception:
            logger.exception("Error updating payment intent quote")
            return

        try:
            await asyncio.wait_for(shutdown_event.wait(), timeout=PUBLISH_INTERVAL_SECONDS)
            return  # shutdown requested
        except TimeoutError:
            pass  # interval elapsed, publish again
